x/evm/evmstate/grpc_query.go

```diff
@@ -438,31 +466,39 @@ func (k Keeper) EstimateGasForEvmCallType(
 		return len(rsp.VmError) > 0, rsp, nil
 	}
 
-	// Execute the binary search and hone in on an executable gas limit
-	hi, err = evm.BinSearch(lo, hi, executable)
+	// BASE CASE:  Jumping straight into binary search is extremely inefficient.
+	// Instead, execute at the highest allowable gas limit first to validate and
+	// set a smarter lower bound.
+	failed, result, err := executable(hi)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("eth call exec error: %w", err)
 	}
-
-	// The gas limit is now the highest gas limit that results in an executable transaction
-	// Reject the transaction as invalid if it still fails at the highest allowance
-	if hi == gasCap {
-		failed, result, err := executable(hi)
-		if err != nil {
-			return nil, fmt.Errorf("eth call exec error: %w", err)
-		}
-
-		if failed && result != nil {
+	// If the base case fails for non-gas reasons, return the error immediately
+	if failed {
+		if result != nil && result.VmError != "" && result.VmError != vm.ErrOutOfGas.Error() {
 			if result.VmError == vm.ErrExecutionReverted.Error() {
 				return nil, fmt.Errorf("estimate gas VMError: %w", evm.NewRevertError(result.Ret))
 			}
-
-			if result.VmError == vm.ErrOutOfGas.Error() {
-				return nil, fmt.Errorf("gas required gas limit (%d)", gasCap)
-			}
-
 			return nil, fmt.Errorf("estimate gas VMError: %s", result.VmError)
 		}
+		return nil, fmt.Errorf("gas required exceeds allowance (%d)", hi)
+	}
+
+	// Set smart lower bound based on actual gas used
+	if result.GasUsed > 0 {
+		lo = result.GasUsed - 1
+	} else {
+		lo = 0
+	}
```

Comment:
this is great, it could transform binary search into a couple of tries only

Reply:
Yeah, it's likely the first run will get the value neeeded